Update page type output with "Update defaults"

trsh opened the pull request:
It all started out here https://github.com/concrete5/concrete5/issues/5365 and ended here https://github.com/concrete5/concrete5/issues/5376. 

Basic idea:  In "Page Types" > "Output" besides the button "Edit Defaults" there is a button "Update defaults". This button sets all child pages blocks to match page type (adds/aliases, removes, reorders, changes parents, etc).

It kind of conflicts with the existing "Setup on Child Pages" idea. But we can discuss and figure it out, I believe.

Mnkras commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,275 @@
+      $this->template = PageTemplate::getByID($pTemplateID);
+
+      if (!$this->template) {
+        throw new \Exception(t('No suchpage template.'));
```
need a space

trsh replied:
Sure

Mnkras commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,275 @@
+            $record = unserialize($message->body);
+            $childPage = \Page::getByID($record['cID'], $record['cvID']);
+
+            var_dump($childPage->getCollectionName());die();
```
no.

trsh replied:
Ouh my gash. Sorry for this :/

Mnkras commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,275 @@
+        $pl = new PageList();
+        $pl->filterByPageTypeID($this->pagetype->getPageTypeID());
+        $pl->filterByPageTemplate($this->template);
+        $pl->ignorePermissions();
```
This can allow people that don't have access to a page to see it. We shouldn't do this.

trsh replied:
Mark, this is only for total results. Same is now in https://github.com/concrete5/concrete5/blob/develop/concrete/controllers/dialog/block/aliasing.php. But you right. I need to add checks for page permissions trough the code.

KorvinSzanto replied:
This is fine

trsh commented:
@Mnkras big TNX. I'll do!

trsh commented:
@Mnkras 

1) Changed spelling issue
2) Removed var_dump.. die
3) As for 

> This can allow people that don't have access to a page to see it. We shouldn't do this.

I think it's OK, because it only for result Count. Same is used in https://github.com/concrete5/concrete5/blob/develop/concrete/controllers/dialog/block/aliasing.php . It's meant that way.

KorvinSzanto commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,273 @@
+          }
+
+          $obj->message = t('All child child pages updated successfully.');
+          echo json_encode($obj);
```
This should be a returned response

KorvinSzanto replied:
If you can update this to use the ResponseFactory and a json response, I think we can merge this.

trsh replied:
@KorvinSzanto tnx and sorry for big delay - I was AFK on my vacation.

- Can you give an example ResponseFactory and json, or point to some docs? Please.

- Why it's OK to use `echo json_encode($obj)` in this script https://github.com/concrete5/concrete5/blob/develop/concrete/controllers/dialog/block/aliasing.php, not in mine :) ?

trsh replied:
@KorvinSzanto I figure it out. But also found one nasty bug, so DO not merge yet, please. I'll fix it soon.

trsh commented:
Don't merge Yet. Found bug while testing. Will fix soon.

trsh commented:
@KorvinSzanto fixed bug and retested.

trsh commented:
@KorvinSzanto , @Mnkras hello!

KorvinSzanto commented:
@aembler can you give this a quick review?

trsh commented:
@aembler can you?

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+              $queue->deleteQueue($name);
+          }
+
+          $obj->message = t('All child child pages updated successfully.');
```
Message contains "child" twice

trsh replied:
:/

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+        }
+
+        $totalItems = $queue->count();
+        \View::element('progress_bar', array('totalItems' => $totalItems, 'totalItemsSummary' => t2("%d pages", "%d pages", $totalItems)));
```
`t2("%d page", "%d pages", ...`)`

trsh replied:
:\

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+
+
+    private function fetchTypeAndTemplate($ptID, $pTemplateID){
+      $cmp = new \Permissions($this->pagetype);
```
Imports at the top

trsh replied:
What imports?

trsh replied:
@a3020 what imports aren't at the top?

a3020 replied:
I'd say, as a best practice, to put your class dependencies (imports) on top. Btw, I'm not the one that will merge / approve your PR. Just giving you feedback on how to improve your PR.

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+
+    public function queueForPageTypeUpdate($pageTypeDefaultPage, $queue)
+    {
+      $records = array();
```
Short array syntax / consistency

trsh replied:
Can do. Don't think it's a problem :/.

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+        $this->set('total', $pl->getTotalResults());
+    }
+
+    public function queueForPageTypeUpdate($pageTypeDefaultPage, $queue)
```
protected?

trsh replied:
OK, yes

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+      $siteTreeID = $site->getSiteTreeID();
+
+      $ptID = $pageTypeDefaultPage->getPageTypeID();
+      $pagesPlusCV = $db->GetAll('select p.cID, max(cvID) as cvID from Pages p inner join CollectionVersions cv on p.cID = cv.cID where ptID = ? and cIsTemplate = 0 and cIsActive = 1 and siteTreeID = ? group by cID order by cID', [$ptID, $siteTreeID]);
```
`fetchAll`

trsh replied:
mhm.. ok.. why?

a3020 replied:
Because `GetAll` is deprecated

trsh replied:
Ok

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+
+        if(count($pageTypeDefaultPageBlocksClone) > 0){
+          foreach ($pageTypeDefaultPageBlocksClone as $pageTypeBlock) {
+            $blocksToAdd[] = array('bID' => $pageTypeBlock->getBlockID(), 'actualDisplayOrder' => $pageTypeBlock->cbDisplayOrder, 'pageTypeArHandle' => $pageTypeBlock->getAreaHandle());
```
Using multiple lines here would improve readibility

trsh replied:
OK

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+      return $queue;
+    }
+
+    public function submit($ptID, $pTemplateID)
```
What happens if you access this URL without these parameters?

trsh replied:
Will try. Good point

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+    public function submit($ptID, $pTemplateID)
+    {
+
+        $responseFactory = $this->app->make(ResponseFactory::class);
```
Do we have access to `$this->app`? https://github.com/concrete5/concrete5/blob/develop/concrete/controllers/backend/user_interface.php and https://github.com/concrete5/concrete5/blob/develop/concrete/src/Controller/Controller.php

trsh replied:
@a3020 I don't get this point and the links. Can write bit more?

a3020 replied:
I'm not sure if the app property exist in this controller. Where is it defined?

mlocati replied:
Every class that extends `Concrete\Core\Controller\AbstractController` has an `$app` property set to the `Concrete\Core\Application\Application` instance.
See this recent discussion: https://concrete5.slack.com/messages/C2N3FLUBW/

trsh replied:
So we can forget this? Besides if $this->app would be empty, I would GOT that. And this was the suggested way from another review.

a3020 replied:
Ah, it has `ApplicationAwareTrait`, ok. Well in that case you could use the container also to e.g. get an instance of the database and replace `Core` with `$this->app` (https://github.com/concrete5/concrete5/pull/5510/files#diff-d8688e8d48fa6ccd325c9f9341b8aa9bR76)

mlocati replied:
> you could use the container also to e.g. get an instance of the database and replace Core with $this->app

:+1:

mlocati replied:
And also, `$db = \Database::connection();` could be written as `$db = $this->app->make(Connection::class);`

trsh replied:
@mlocati, @a3020 and why? What's the problem?

mlocati replied:
Speed: using $app->make is faster that using facades (`\Core::make`, `\Database::connection`, ...)

trsh replied:
@mlocati ok. I like speed. But how far I go with this? 

\Permissions **to** $p = $this->app->make(Permissions ::class);
throw new \Exception(t('You do not have access to edit this page type.')); **to** throw new $this->app->make(Exception ::class)(t('You do not have access to edit this page type.'));

??

Then its \Queue and \stdClass, \Page, \Block o_O

trsh replied:
@mlocati mark that this does not effect some page load speed, it's a tool Editors will use maybe once / twice day. Saving some milis here is not Critical.  Critical is the fact, that this tool doesn't exist :)

mlocati replied:
> ok. I like speed. But how far I go with this?

As long as your patience lasts :wink:

IMHO that doesn't make sense for PHP objects (like `Exception`), but using if for concrete5 objects grants a greater level of extensibility (we can add dependencies to constructors as we need them) and speed (we avoid facades).

mlocati replied:
> mark that this does not effect some page load speed, it's a tool Editors will use maybe once / twice day.

Yes, you're right here. BTW we already have an `Application` instance: typing `$this->app->make` instead of `Core::make` only costs you 6 chars to be typed.
Avoiding other facades is only a general approach I personally always follow: I don't think "this method doesn't require speed so use facades, that method requires speed so use $app". But of course, you are free to do whatever you like :wink:

trsh replied:
@mlocati ok, thanks for the info. I'm a little bit mad, because this P.R. has been postponed because of "nice to have" features for months. And the code I took examples of and is currently in production has all issues Your pointing out here... Will correct anyways!

trsh replied:
I mean there is a big probability, that after these fixes, somebody will look at my code again in middle of September, and write down some spelling mistake.. and then it will be November and so on :D. (partly a joke)

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+
+              $actions = $blockToUpdate['actions'];
+
+              foreach ($actions as $action) {
```
Decrease levels of indentation? Simplify logic? Extracting to methods / classes?

trsh replied:
You suggest to serialize a Class? Because queue->send is string and I must use it.

trsh replied:
@a3020  ?

a3020 replied:
Not sure where I mentioned to 'serialize a class'? I meant that you have a lot of logic here that's is completely undocumented and hard to follow. It would make perfect sense to extract this big chunk of code to methods and optionally add comments. That would make it easier for other devs to read the code, and e.g. make changes to it.

trsh replied:
@a3020 ok I will see what can I do. Anyways this is already huge step-up when comparing to code, I used as an example > https://github.com/concrete5/concrete5/blob/develop/concrete/controllers/dialog/block/aliasing.php

a3020 replied:
If it gets merged in as it is, other developers will take your code as an example. But again, I'm not the one to approve, just giving some feedback. 😉

trsh replied:
@a3020 can you contact approval guys to take a look?

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+                  }else{
+                    $pageTypeBlock->alias($childPage);
+                  }
+                }
```
Running the php-cs-fixer would solve some code styling issues here.

trsh replied:
Jahu

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+                }
+                else if ($action['name'] == 'change_arHandle') {
+
+                  $nh = $action['actualArHandle'];
```
Unclear variable names, at least, to me.

trsh replied:
OK

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+
+          $obj->totalItems = $queue->count();
+
+          if ($queue->count() == 0) {
```
`===`

trsh replied:
OK

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+        $request = $this->request;
+        $ptID = $request->get('ptID');
+
+        if (!$ptID || !$request->get('ptID')) {
```
Is there a reaseon you test for `ptID`, but not for `pTemplateID`?

trsh replied:
No! this is stupid. Will fix

a3020 commented on concrete/controllers/dialog/type/update_from_type.php:
```diff
@@ -0,0 +1,284 @@
+    protected $viewPath = '/dialogs/type/update_from_type';
+
+
+    private function fetchTypeAndTemplate($ptID, $pTemplateID){
```
Move `private` methods to the bottom?

trsh replied:
Ok

trsh commented:
@a3020 thanks. Will fix soon!

trsh commented:
@a3020 I have taken _most_ of your comments into account and pushed changes. I do not understand the problem with "continuous-integration/appveyor/pr". How can we prosseed?

trsh commented:
@mlocati, @KorvinSzanto, @a3020, @Mnkras any ideas how to take this to the next level? Days are again going by one by another.

mlocati commented:
@trsh I'm just a community member like you...

trsh commented:
@mlocati OK. Maybe u know how to ping core team?

mlocati commented:
They should already ve notified about these comments. I think that, since this PR is quite big, you'll have to wait until they have some spare time...

aembler commented:
We are indeed quite busy.

The updates that you're proposing are pretty expansive, and I'm glad that others have been able to do some of the interim vetting and reviewing of the pull request in a timely fashion. Large changes that are unsolicited are going to have a much longer route to getting into the core than things that we are explicitly looking for – things marked "ready to start" and especially "like/love to have." 

Additionally, this pull request also had the bad luck of coming in right as we were readying a major release, and it clearly wasn't mature at the time to be included in 8.2. 8.2.1 was an immediate maintenance release, so this wasn't going to make it into that either. 

We are committed to reviewing this pull request, and will do so when we have time.

aembler commented:
Thanks for your patience. I have made some changes to this but they were minor. This will be a nice addition.

MrKarlDilkington commented:
@trsh 

Can you explain the full list of changes that this pull request makes, please. This explanation will be used to update the concrete5 documentation.

trsh commented:
@MrKarlDilkington 

Actually is less about change, and more about new addition.

> Basic idea: In "Page Types" > "Output" besides the button "Edit Defaults" there is a button "Update defaults". This button sets all child pages blocks to match page type (adds/aliases, removes, reorders, changes parents, etc).

The lines above do not help? If not, please show me some example how to form it :)

@aembler 

Tnx

MrKarlDilkington commented:
@trsh 

> This button sets all child pages blocks to match page type (adds/aliases, removes, reorders, changes parents, etc).

Can you elaborate on the "etc." part, please.

mlocati commented:
In PHP 5.x we can't have this code:
`$this->app->make(Block::class)::getByID(...)`.

#5948 (and in particular https://github.com/concrete5/concrete5/pull/5948/commits/639512b4afe22b8131b7e9d1fb6c86ff493a5372) fixes it.

trsh commented:
@MrKarlDilkington 

I think we can just drop the "(...)" part.

trsh commented:
@mlocati Great!
